src/concurrency/sync.rs

```diff
@@ -354,15 +366,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         mutex.clock.clone_from(clock)
                     });
                 }
-                if let Some(thread) = this.machine.sync.mutexes[id].queue.pop_front() {
-                    this.unblock_thread(thread, BlockReason::Mutex(id))?;
+                let thread_id = mutex.queue.pop_front();
+                if thread_id.is_some() {
+                    // We need to drop mutex before unblock_thread because mutex will be borrowed again
+                    // in the unblock callback.
```

Comment:
```suggestion
                    // We need to drop mutex before unblock_thread because mutex will be borrowed again
                    // in the unblock callback.
```
```suggestion
                    // We need to drop our mutex borrow before unblock_thread
                    //  because it will be borrowed again in the unblock callback.
```